### DISCUSSION TITLE
v1.0.0

### DISCUSSION THREAD

**Discussion by MilesCranmer** [MilesCranmer/PySR#755] [General]
# PySR v1.0.0 Release Notes

PySR 1.0.0 introduces new features for imposing specific functional forms and finding parametric expressions. It also includes TensorBoard support, along with significant updates to the core algorithm, including some important bug fixes. The default hyperparameters have also been updated based on extensive tuning, with a maxsize of 30 rather than 20.

### Expression Specifications

PySR 1.0.0 introduces new ways to specify the structure of equations through "Expression Specifications", that expose the new backend feature of `AbstractExpression`:

#### Template Expressions
`TemplateExpressionSpec` allows you to define a specific structure for your equations. For example:

```python
expression_spec = TemplateExpressionSpec(
    function_symbols=["f", "g"],
    combine="((; f, g), (x1, x2, x3)) -> sin(f(x1, x2)) + g(x3)"
)
```

#### Parametric Expressions
`ParametricExpressionSpec` enables fitting expressions that can adapt to different categories of data with per-category parameters:

```python
expression_spec = ParametricExpressionSpec(max_parameters=2)
model = PySRRegressor(
    expression_spec=expression_spec
    binary_operators=["+", "*", "-", "/"],
)
model.fit(X, y, category=category)  # Pass category labels
```

### Improved Logging with TensorBoard

The new `TensorBoardLoggerSpec` enables logging of the search process, as well as hyperparameter recording, which exposes the `AbstractSRLogger` feature of the backend:

```python
logger_spec = TensorBoardLoggerSpec(
    log_dir="logs/run",
    log_interval=10,  # Log every 10 iterations
)
model = PySRRegressor(logger_spec=logger_spec)
```

Features logged include:

- Loss curves over time at each complexity level
- Population statistics
- Pareto "volume" logging (measures performance over all complexities with a single scalar)
- The min loss over time

### Updated Default Parameters

The default hyperparameters have been significantly revised based on testing:

- Increased default `maxsize` from 20 to 30, as I noticed that many people use the defaults, and this maxsize would allow for more accurate expressions.
- New mutation operator weights optimized for better performance, along the new mutation "rotate tree."
- Improved search parameters tuned using Pareto front volume calculations.
- Default `niterations` increased from 40 to 100, also to support better accuracy (at the expense of slightly longer default search times).

### Core Changes

- New output organization: Results are now stored in `outputs/<run_id>/` rather than in the directory of execution.
- Improved performance with better parallelism handling
- Support for Python 3.10+ 
- Updated Julia backend to version 1.10+
- Fix for aliasing issues in crossover operations

## Breaking Changes

- Minimum Python version is now 3.10
- Output file structure has changed to use directories
- Parameter name updates:
  - `equation_file` → `output_directory` + `run_id`
  - Added clearer naming for parallelism options, such as `parallelism="serial"` rather than the old `multithreading=False, procs=0` which was unclear

## Documentation

The documentation has a new home at https://ai.damtp.cam.ac.uk/pysr/

<hr /><em>This discussion was created from the release <a href='https://github.com/MilesCranmer/PySR/releases/tag/v1.0.0'>v1.0.0</a>.</em>

**Comment by GoldenGoldy** [comment]
Before I forget: congrats on releasing the new version! Great new features!

Might it be though that v1.0.0 is not completely stable yet? I'm asking because I already had a job crash twice since the update, while I did not have such problems before. I didn't open a bug report for this yet because it's a bit difficult to say for sure whether it's due to PySR itself or Julia or something else.

One job crashed with this message, after about 5 hours:

> juliacall.JuliaError: TaskFailedException
Stacktrace:
  [1] wait(t::Task)
    @ Base ./task.jl:370
  [2] fetch
    @ ./task.jl:390 [inlined]
    ........................
        nested task error: On worker 3:
    DomainError with Inf:
    sin(x) is only defined for finite x.
    Stacktrace:
      [1] sin_domain_error
        @ ./special/trig.jl:28
        ....................

And a second job crashed with the following message, after about 11 hours:

> juliacall.JuliaError: TaskFailedException
Stacktrace:
  [1] wait(t::Task)
    @ Base ./task.jl:370
  [2] fetch
    @ ./task.jl:390 [inlined]
    .........................
        nested task error: Distributed.ProcessExitedException(29)
    Stacktrace:
      [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
        @ Base ./task.jl:958
      [2] wait()
        @ Base ./task.jl:1022
        .......................
        Worker 29 terminated.
UNHANDLED TASK ERROR: IOError: read: connection reset by peer (ECONNRESET)
Stacktrace:
  [1] wait_readnb(x::Sockets.TCPSocket, nb::Int64)
    @ Base ./stream.jl:410
  [2] (::Base.var"#wait_locked#832")(s::Sockets.TCPSocket, buf::IOBuffer, nb::Int64)
    @ Base ./stream.jl:972
    ......................

In both cases I was running in distributed mode (cluster_manager='slurm'), with 30 CPU cores. The Python script is the same as I was using with previous PySR versions (apart from required changes such as passing the parallelism parameter), when I didn't have these issues. Might it be due to the PySR update that these issues now occur? Or should I look at the environment? (Julia version and so on)

**Reply by MilesCranmer** [reply]
Do you have the full traceback? And what is your script? Please post an issue if possible. These errors don't look related to the environment but something else.

**Reply by GoldenGoldy** [reply]
Sure! Issue created here: https://github.com/MilesCranmer/PySR/issues/759
I tried to provide as much detail as possible.
